lib/parser/source/tree_rewriter.rb

```diff
@@ -326,6 +318,24 @@ def transaction
         @in_transaction = previous
       end
 
+      ##
+      # Provides access to a diagnostic engine.
+      # By default outputs diagnostic to $stderr
+      #
+      def self.default_diagnostics
+        @default_diagnostics ||= Diagnostic::Engine.new.tap do |engine|
+          engine.consumer = -> diag { $stderr.puts diag.render }
+        end
+      end
+
+      ##
+      # Provides access to a diagnostic engine.
+      # By default: self.class.default_diagnostics
+      #
+      def diagnostics
+        @diagnostics ||= self.class.default_diagnostics.dup
```

Comment:
Do you need `.dup`? Engine has no state except configuration, so re-using the same instance seems to be OK.

Reply:
It has state, which is `consumer`...

Reply:
`.dup` makes a copy of the top-level wrapper:
```
> irb
2.7.1 :001 > o1 = Object.new; o2 = Object.new; o1.instance_variable_set(:@ivar, o2); o1
 => #<Object:0x00007fda76292a90 @ivar=#<Object:0x00007fda76292a68>>
2.7.1 :002 > o1.dup
 => #<Object:0x00007fda7319b2e8 @ivar=#<Object:0x00007fda76292a68>>
```

^ `@ivar` is the same (and this why Ruby has `initialize_copy` hook - to make a deep copy if you need it)

Even if `consumer` is something stateful it's not copied this way. However, if you have a custom diagnostic engine that has this hook (and thus, makes a "deep" copy) then this `.dup` is valuable.

I'm still not sure about it (because you can't put `.dup` everywhere to protect users from re-using cached objects). If users want to override something they should add this `.dup` manually. Ah, let's keep it, at least this way nobody needs to know that objects are cached and re-used (while performance doesn't really suffer)

Reply:
The issue is that if a use creates a `TreeRewriter`, then does `my_rewriter.diagnostics.consumer = -> { ... }`, the *default* consumer has now changed for all future rewriters.
I'm working on a modified PR that changes the API a bit and might be better and cleaner, just haven't had the time to polish it yet. Stay tuned :-)